### ISSUE TITLE
ctypeslib.as_ctypes cannot wrap large arrays (> 2 ** 31 bytes)

### ISSUE THREAD

**Issue by miketrumpis** [numpy/numpy#14214]
After updating Numpy I can no longer use `numpy.ctypeslib.as_ctypes` for an array whose size is larger than 2 ** 31 bytes (a signed C int). This appears to be due to the changes from 08a92de354f1f7592c7957f4aafe892f3d5a828a that allow more general ctypes pointers.

Old style pointer:
https://github.com/numpy/numpy/blob/971e2e89d08deeae0139d3011d15646fdac13c92/numpy/ctypeslib.py#L402-L404

New style pointer:
https://github.com/numpy/numpy/blob/5a5987291dc95376bb098be8d8e5391e89e77a2c/numpy/ctypeslib.py#L538-L540

I'm using ctypeslib in the context of making large `multiprocessing.sharedctypes.Array` objects available to multiprocessing pools. 

I would wager that this C int size limit runs deep, making a direct fix difficult. I could simply monkey-patch to previous functionality, since I know that my arrays work with the previous, more restrictive wrapping. But I think splitting large (> 2 GB) arrays across pool workers is probably a common use case (I hope I didn't miss an existing issue). 

I'm a bit out of depth, but would it make sense to fall back to the old behavior after testing for no-strides and native endian-ness?

### Reproducing code example:

```python
import numpy as np
arr = np.zeros((2 ** 31 + 1,), 'b')
assert(arr.size * arr.itemsize > 2 ** 31)
c_arr = np.ctypeslib.as_ctypes(arr)      
```

Under versions > 1.16.0 I am getting an error (traceback below), because the dtype memory footprint doesn't fit into a signed C int. 

### Error message:

```python                                                                                                            
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-e91c3f6f3634> in <module>
----> 1 c_arr = np.ctypeslib.as_ctypes(arr)

~/.pyenv/versions/3.6.4/envs/ctype_test/lib/python3.6/site-packages/numpy/ctypeslib.py in as_ctypes(obj)
    536             raise TypeError("readonly arrays unsupported")
    537 
--> 538         dtype = _dtype((ai["typestr"], ai["shape"]))
    539         result = as_ctypes_type(dtype).from_address(addr)
    540         result.__keep = obj

ValueError: invalid shape in fixed-type tuple: dtype size in bytes must fit into a C int.
```

### Numpy/Python version information:
I down-versioned numpy until `as_ctypes` worked again:

Result of `import sys, numpy; print(numpy.__version__, sys.version)`

*Where it works*
```
1.16.0 3.6.4 (default, Feb 22 2018, 18:37:35) 
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```

*Where it breaks*
```
1.17.0 3.6.4 (default, Feb 22 2018, 18:37:35) 
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```

**Comment by eric-wieser** [comment]
Darn, that's an unfortunate side effect - I'd forgotten that the maximum dtype size is less than the maximum array size.

**Comment by eric-wieser** [comment]
Replacing those lines with the following should fix it
```
# can't use `_dtype((ai["typestr"], ai["shape"]))` here, as it hits the dtype size limit (gh-14214)
ctype_scalar = as_ctypes_type(ai["typestr"])
result = _ctype_ndarray(ctype_scalar , shape)
```
No need for any fallback behavior

**Comment by miketrumpis** [comment]
I made this change to the v1.17.0 tag and it looks like I'm getting equal results as from 1.16.0 for some multiprocessing number crunching. For my purposes, it seems fine. Does it not break support for structured arrays, etc?

```python
ctype_scalar = as_ctypes_type(ai["typestr"])
result = _ctype_ndarray(ctype_scalar, ai["shape"]).from_address(addr)
result.__keep = obj
return result
```

**Comment by miketrumpis** [comment]
Also numpy.test() passes.

**Comment by eric-wieser** [comment]
As far as I'm aware, it's functionally a strictly better version is every way. I'd argue the first formulation was a little cleaner, but that's irrelevant given it didn't always work

**Comment by charris** [comment]
The fix is needed both for 1.16.x and 1.17.x
